diff --git a/openmaptiles/imposm.py b/openmaptiles/imposm.py
index 912b2cb..d90aa7f 100644
--- a/openmaptiles/imposm.py
+++ b/openmaptiles/imposm.py
@@ -1,8 +1,6 @@
 import re
 from .tileset import Tileset
 
-re_zoom_level = re.compile("^.*_z(\d+)$")
-
 
 def zres(pixel_scale, zoom):
     # See https://github.com/openmaptiles/postgis-vt-util/blob/master/src/ZRes.sql
@@ -14,13 +12,6 @@ def call_zres(pixel_scale, match):
     return str(zres(pixel_scale, match.group(0)[4:6]))
 
 
-def get_zoom_level(table_name):
-    m = re_zoom_level.match(table_name)
-    if m is not None:
-        return int(m.groups()[0])
-    return 0
-
-
 def create_imposm3_mapping(tileset_filename):
     tileset = Tileset.parse(tileset_filename)
     definition = tileset.definition
@@ -40,9 +31,7 @@ def create_imposm3_mapping(tileset_filename):
 
     for layer in tileset.layers:
         for mapping in layer.imposm_mappings:
-            layer_mapping = mapping.get('generalized_tables', {})
-            for table_name in sorted(layer_mapping.keys(), key=lambda x: get_zoom_level(x), reverse=True):
-                definition = layer_mapping[table_name]
+            for table_name, definition in mapping.get('generalized_tables', {}).items():
                 if 'tolerance' in definition:
                     try:  # Test if numeric
                         float(definition['tolerance'])
@@ -78,6 +67,6 @@ def create_imposm3_mapping(tileset_filename):
 
     return {
         'tags': dict(include=list(sorted(set(include_tags)))),
-        'tables': tables,
         'generalized_tables': generalized_tables,
+        'tables': tables,
     }
